### ISSUE TITLE
IterableDataset skipping lines in CSV file?

### ISSUE THREAD

**Issue by tb438** [pytorch/pytorch#31188]
## 🐛 Bug

I am writing a dataloader pipeline for PyTorch and It seems that there is a bug in the IterableDataset object.

## To Reproduce

```
`#!/usr/bin/env python3

import random
import time
import torch
import DataLoader as loader
from itertools import cycle, islice
from torch.utils.data import Dataset, IterableDataset, DataLoader

class MyIterableDataset(IterableDataset):

    def __init__(self, file_path):
        self.file_path = file_path

    def parse_file(self, file_path):
        with open(file_path, 'r') as file_obj:
            for line in file_obj:
                tokens = line.strip('\n').split(' ')
                yield from tokens

    def get_stream(self, file_path):
        return cycle(self.parse_file(file_path))

    def __iter__(self):
        return self.get_stream(self.file_path)

def irandomize(iterable):
    iterable = iter(iterable)
    items = []
    try:
        while True:
            for i in range(random.randint(1, 10)):
                items.append(iterable.next())
            random.shuffle(items)
            for i in range(random.randint(1, 10)):
                if items:
                    yield items.pop()
                else:
                    break
    except StopIteration:
        random.shuffle(items)
        for item in items:
            yield item
        raise StopIteration

if __name__ == '__main__':

    dataset = MyIterableDataset('dataset.csv')

    loader = irandomize(DataLoader(dataset, batch_size = 1))

    for batch in islice(loader, 64):
        print(batch)
    for batch in islice(loader, 64):
        print(batch)

$ python3 dataloader.py | awk -F',' '{print $1}' | cut -c3-10 | sort -n
/usr/lib/python3/dist-packages/sklearn/externals/joblib.py:1: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
...
103
104
105
106
107
108
110
111
112
113
114
115
116
117
118
119
120
121
122
124
125
126
127
128
129
133
```
## Expected behavior

Both range sizes are to fetch 64 rows, but for some reason rows 109 and 123 are getting skipped with the DataLoader object?
 
## Environment

Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 19.10
GCC version: (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1060 6GB
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.3.1
[pip3] torchgan==0.0.3
[pip3] torchvision==0.4.2
[conda] Could not collect

cc @SsnL

**Comment by tb438** [comment]
Hmm it looks like there is something wrong with my irandomize() function; whenever I don't shuffle the iterable, it returns the correct number and sequence of rows.

So with IterableDataset not supporting a shuffle operation, what is the best way of streaming (and shuffling) rows from a very large CSV file? It doesn't look like IterableDataset supports a length argument, so my thoughts were to first wc -l the CSV file, then use that as the total number of rows in the dataset that multiple workers should be dispatched across?

**Comment by colesbury** [comment]
My suggestion is: don't use IterableDataset and use `shuffle=True` on your DataLoader.

1) Load all your data into a Python list
2) Use Dataset
3) Use shuffle=True in the consturtor to DataLoader

If you want to cycle and make the dataset really large, override `__len__` and `__getitem__` in your dataset.

**Comment by tb438** [comment]
I don't think this will work, my dataset is 15G and when using a conventional DataLoader it locks everything up (presumably because the OS is swapping). Does the regular DataLoader support lazy loads now? Or is that only available through the IterableDataset object? IterableDataset object doesn't yet support len() or shuffle ops yet unfortunately, but it does lazy load the entire 15G .csv without locking the whole system up.

**Comment by SsnL** [comment]
Question to your original problem: Isn't this expected? You are only taking 128 batches. And at the end of the second `islice(*, 64)`, the current buffer is shuffled and yielded, which means that part of it could be before the cutoff (idx=64) and part of it could be after.. It seems that 109 and 123 are just after the cutoff.
